samza-api/src/main/java/org/apache/samza/operators/windows/Windows.java

```diff
@@ -119,11 +121,21 @@ private Windows() { }
    * @param <K> the type of the key in the {@link Window}
    * @return the created {@link Window} function.
    */
-  public static <M, K, WV> Window<M, K, WV> keyedTumblingWindow(Function<M, K> keyFn, Duration interval,
-                                                                Supplier<WV> initialValue, FoldLeftFunction<M, WV> foldFn) {
+  public static <M, K, WV> Window<M, K, WV> keyedTumblingWindow(Function<? super M, ? extends K> keyFn, Duration interval,
+                                                                Supplier<? extends WV> initialValue, FoldLeftFunction<? super M, WV> foldFn) {
 
     Trigger<M> defaultTrigger = new TimeTrigger<>(interval);
-    return new WindowInternal<M, K, WV>(defaultTrigger, initialValue, foldFn, keyFn, null, WindowType.TUMBLING);
+    return new WindowInternal<>(defaultTrigger, initialValue::get, new FoldLeftFunction<M, WV>() {
```

Comment:
This is to formulate the type of Function that our internal class will depend on. Leaking the captured type parameter to internal functions leads to a whole lot changes over the full implementation stack, which does not seem to be necessary for the purpose (i.e. just allow relaxed type parameter from user-supplied functions).